### DISCUSSION TITLE
Should `finish` be added to nvim-dap?

### DISCUSSION THREAD

**Discussion by mwcz** [mfussenegger/nvim-dap#1178] [Ideas]
Hi!  I'm a heavy user of nvim-dap and nvim-dap-ui, with codelldb for debugging Rust.  Rust makes heavy use of implicit return values, like this super contrived example:

```rust
/* 1 */  fn my_kinda_rand() -> f32 {
/* 2 */    (1.0 + rand()) / 2.0
/* 3 */  }
```

My experience with nvim-dap is that it's not possible to put a breakpoint on line 2 and inspect the value.  To get around that, I frequently bind the value to a variable, which I can then inspect with the debugger.

```rust
/* 1 */  fn my_kinda_rand() -> f32 {
/* 2 */    let r = (1.0 + rand()) / 2.0;
/* 3 */    r
/* 4 */  }
```

The problem is that this leads to a lot of non-idiomatic Rust, and clippy (Rust's linter) [doesn't like it](https://rust-lang.github.io/rust-clippy/master/index.html#/let_and_return).  It's also harder to take this approach when debugging third party code.

Codelldb and gdb both have `finish`, which runs the current stack frame to completion, and prints the return value.

![image](https://github.com/mfussenegger/nvim-dap/assets/364615/9ba50c4d-7f8d-4713-981f-c33134756cae)

Perfect!  The only hangup is that I haven't found a way to surface that return value in nvim-dap.  Is this a feature that others would find useful?  Is there already a way to do it that I missed?  Thanks for the discussion.

**Comment by mfussenegger** [comment]
This relates a bit to https://github.com/microsoft/debug-adapter-protocol/issues/458

There's currently nothing in the specification that would allow nvim-dap to surface this in a way that works across debug adapters.

My workaround so far is to just select the whole expression and then use the hover widget on it

**Reply by mwcz** [reply]
I see, :crossed_fingers: the dap proposal is accepted, and thank you for the hover tip, that will help a ton in the meantime.

**Reply by mwcz** [reply]
It seems that's closed now, but I don't grok what the resolution was.  I see mention of a `returnValue` added after `stepOut`, which would do the trick I suppose.
